File: Dynamic_Programming/1335. Minimum Difficulty of a Job Schedule.cpp
/*
1335. Minimum Difficulty of a Job Schedule
*/

// TC:O(n^2*d) SC:O(n*d)
class Solution {
public:
    int minDifficulty(vector<int>& jobDifficulty, int d) {
        int n = jobDifficulty.size();
        if(n < d) return -1;
        jobDifficulty.insert(jobDifficulty.begin(), 0);
        vector<vector<int>> dp(n+1, vector<int>(d+1, INT_MAX/2));
        dp[0][0] = 0;

        for(int i = 1; i <= n; i++){
            for(int k = 1; k <= min(d,i); k++){
                //要保留k-1個給前面的分 因為每組至少要有一個
                int maxVal = jobDifficulty[i];
                for(int j = i; j >= k; j--){
                    maxVal = max(maxVal, jobDifficulty[j]);
                    dp[i][k] = min(dp[i][k], dp[j-1][k-1] + maxVal);
                }
            }
        }

        return dp[n][d];

    }
};


/*
分割出d天 每天至少一個任務 每天的數值取決於當天最大值 如何最小化各個subarray的總和是多少
原本想說能否貪心的去做 第一組肯定取最多個元素 那一組的區間會是最大的 看他能怎麼取將一些較大元素包在同個區間內 剩下的自成一國
但很難去判定是要從哪個元素開始貪心的取 分割subarray並且互斥->區間型DP嘗試
區間型I dp[i][k] = dp[j-1][k-1] + max{j:i} 
given an array s[0:i], minimize the sum of each maximum of k subarray
*/

File: Dynamic_Programming/1335. Minimum Difficulty of a Job Schedule.py
# TC:O(n*d) SC:O(n*d) 可在空間壓縮到O(n)
class Solution:
    def minDifficulty(self, jobDifficulty: List[int], d: int) -> int:
        n = len(jobDifficulty)
        if n < d: return -1
        # 單調棧優化
        f = [[inf] * n for _ in range(d)]
        f[0] = list(accumulate(jobDifficulty, max))
        for i in range(1, d):
            st = [] # stack
            for j in range(i, n):
                mn = f[i-1][j-1]
                while st and jobDifficulty[st[-1][0]] <= jobDifficulty[j]:
                   mn = min(mn, st.pop()[1])
                f[i][j] = mn + jobDifficulty[j] # 以他為最大值的狀況
                if st: # 如果還有剩 那他就有可能是被之前包含的狀況會更小 
                    f[i][j] = min(f[i][j], f[i][st[-1][0]])
                st.append((j, mn))
        return f[d-1][n-1]

        '''
        # 空間優化 該種定義方式才這麼容易做到
        f = list(accumulate(jobDifficulty, max))
        for i in range(1, d):
             # 這裡可以從值域去想 當下的數值想發揮影響只有在他是當前最大值的狀況下
             # 並且如果他走在以他為最大的數值內 他需要去找到最小的 切割點 就會是包含他為最大值的狀況下的最佳切割策略
            for j in range(n-1, i-1, -1):
                f[j] = inf
                mx = 0
                for k in range(j, i - 1, -1):
                    if jobDifficulty[k] > mx: mx = jobDifficulty[k]
                    if f[k - 1] + mx < f[j]: f[j] = f[k - 1] + mx
        return f[-1]
        '''
        '''
        遞推
        f = [[inf] * n for _ in range(d)]
        f[0] = list(accumulate(jobDifficulty, max))
        for i in range(1, d):
            for j in range(i, n):
                mx = 0
                for k in range(j, i - 1, -1):
                    if mx < jobDifficulty[k]: mx = jobDifficulty[k]
                    if f[i - 1][k - 1] + mx < f[i][j]: f[i][j] = f[i - 1][k - 1] + mx
        return f[-1][-1]
        '''

        '''
        遞歸
        @cache
        def dfs(cnt:int, i:int) -> int:
            if cnt == 0:
                return max(jobDifficulty[:i+1])
            
            mx, ans = 0, inf
            for j in range(i, cnt-1, -1):
                if jobDifficulty[j] > mx:
                    mx = jobDifficulty[j]
                tmp = mx + dfs(cnt-1, j-1)
                if tmp < ans:
                    ans = tmp
            return ans
        return dfs(d-1, n-1)
        '''
        

# 不同思考方向 就走不到 單調棧與空間優化
class Solution:
    def minDifficulty(self, jobDifficulty: List[int], d: int) -> int:
        n = len(jobDifficulty)
        if n < d: return -1
        
        f = [[inf] * n for _ in range(d)]
        f[0] = list(accumulate(jobDifficulty, max))
        for i in range(1, d):
            for j in range(i, n):
                mx = 0
                for k in range(j, i - 1, -1):
                    mx = max(mx, jobDifficulty[k])
                    f[i][j] = min(f[i][j], f[i - 1][k - 1] + mx)
        return f[-1][-1]

        '''
        遞推
        @cache
        def dfs(i:int, cnt:int) -> int:
            if cnt == 0:
                return max(jobDifficulty[:i+1])
            
            mx, ans = 0, inf
            for j in range(i, cnt-1, -1):
                if jobDifficulty[j] > mx:
                    mx = jobDifficulty[j]
                tmp = mx + dfs(j-1, cnt-1)
                if tmp < ans:
                    ans = tmp
            return ans
        return dfs(n-1, d-1)
        '''
        
